federationapi/routing/join.go

```diff
@@ -33,153 +34,144 @@ import (
-		}
-	}
-	if !inRoomRes.IsInRoom {
-		return util.JSONResponse{
-			Code: http.StatusNotFound,
-			JSON: spec.NotFound(fmt.Sprintf("Room ID %q has no remaining users on this server", roomID)),
+		stateEvents := make([]gomatrixserverlib.PDU, len(queryRes.StateEvents))
+		for i, stateEvent := range queryRes.StateEvents {
+			stateEvents[i] = stateEvent.PDU
 		}
+		return event, stateEvents, nil
 	}
 
-	// Check if the restricted join is allowed. If the room doesn't
-	// support restricted joins then this is effectively a no-op.
-	res, authorisedVia, err := checkRestrictedJoin(httpReq, rsAPI, roomVersion, roomID, userID)
-	if err != nil {
-		util.GetLogger(httpReq.Context()).WithError(err).Error("checkRestrictedJoin failed")
-		return spec.InternalServerError()
-	} else if res != nil {
-		return *res
+	roomQuerier := JoinRoomQuerier{
+		roomserver: rsAPI,
 	}
 
-	// Try building an event for the server
-	proto := gomatrixserverlib.ProtoEvent{
-		Sender:   userID,
-		RoomID:   roomID,
-		Type:     "m.room.member",
-		StateKey: &userID,
+	input := gomatrixserverlib.HandleMakeJoinInput{
+		Context:            httpReq.Context(),
+		UserID:             userID,
+		RoomID:             roomID,
+		RoomVersion:        roomVersion,
+		RemoteVersions:     remoteVersions,
+		RequestOrigin:      request.Origin(),
+		RequestDestination: request.Destination(),
+		LocalServerName:    cfg.Matrix.ServerName,
+		RoomQuerier:        &roomQuerier,
+		BuildEventTemplate: createJoinTemplate,
 	}
-	content := gomatrixserverlib.MemberContent{
-		Membership:    spec.Join,
-		AuthorisedVia: authorisedVia,
-	}
-	if err = proto.SetContent(content); err != nil {
-		util.GetLogger(httpReq.Context()).WithError(err).Error("builder.SetContent failed")
-		return spec.InternalServerError()
+	response, internalErr := gomatrixserverlib.HandleMakeJoin(input)
+	if internalErr != nil {
+		return *internalErr
 	}
 
-	identity, err := cfg.Matrix.SigningIdentityFor(request.Destination())
-	if err != nil {
-		return util.JSONResponse{
-			Code: http.StatusNotFound,
-			JSON: spec.NotFound(
-				fmt.Sprintf("Server name %q does not exist", request.Destination()),
-			),
-		}
-	}
-
-	queryRes := api.QueryLatestEventsAndStateResponse{
-		RoomVersion: roomVersion,
-	}
-	event, err := eventutil.QueryAndBuildEvent(httpReq.Context(), &proto, cfg.Matrix, identity, time.Now(), rsAPI, &queryRes)
-	if err == eventutil.ErrRoomNoExists {
-		return util.JSONResponse{
-			Code: http.StatusNotFound,
-			JSON: spec.NotFound("Room does not exist"),
-		}
-	} else if e, ok := err.(gomatrixserverlib.BadJSONError); ok {
-		return util.JSONResponse{
-			Code: http.StatusBadRequest,
-			JSON: spec.BadJSON(e.Error()),
-		}
-	} else if err != nil {
-		util.GetLogger(httpReq.Context()).WithError(err).Error("eventutil.BuildEvent failed")
+	if response == nil {
+		util.GetLogger(httpReq.Context()).Error("gmsl.HandleMakeJoin returned invalid response")
```

Comment:
I strongly dislike this: we shouldn't be returning nil responses AND nil errors. If I see this log line in the logs it gives me no clue what went wrong here. Please return an error instead with contextual information. Returning `nil, nil` makes sense only when A) the data being `nil` isn't an error and B) the `nil` indicates absence rather than as a result of an `OR` operation (error OR data).

Reply:
We don't ever actually return `nil, nil`, but since it returns a pointer to the result it technically could have been nil.
The result is a pointer since it returns a full `ProtoEvent` struct. I could change that to be behind a pointer, but that could also technically be nil too. 
Which we could just not check, and just let the server panic assuming it was a programmer error that led to this case. I really don't like leaving in intentional panics across a library boundary. It's one thing to have them for internal programming errors, but relying on an external library api behaviour is dangerous.